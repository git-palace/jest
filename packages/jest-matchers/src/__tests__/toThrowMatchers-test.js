/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+jsinfra
 */

'use strict';

const jestExpect = require('../').expect;
const matchErrorSnapshot = require('./_matchErrorSnapshot');

['toThrowError', 'toThrow'].forEach(toThrow => {
  describe('.' + toThrow + '()', () => {

    class Err extends Error {}
    class Err2 extends Error {}

    test('to throw or not to throw', () => {
      jestExpect(() => { throw new Error('apple'); })[toThrow]();
      jestExpect(() => {}).not[toThrow]();
    });

    describe('strings', () => {
      it('passes', () => {
        jestExpect(() => { throw new Error('apple'); })[toThrow]('apple');
        jestExpect(() => { throw new Error('banana'); })
          .not[toThrow]('apple');
        jestExpect(() => {}).not[toThrow]('apple');
      });

      test('did not throw at all', () => {
        matchErrorSnapshot(() => jestExpect(() => {})[toThrow]('apple'));
      });

      test('threw, but message did not match', () => {
        matchErrorSnapshot(() =>
          jestExpect(() => { throw new Error('apple'); })
            [toThrow]('banana'),
        );
      });

      it('properly escapes strings when matching against errors', () => {
        jestExpect(() => { throw new TypeError('"this"? throws.'); })
          [toThrow]('"this"? throws.');
      });

      test('threw, but should not have', () => {
        matchErrorSnapshot(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow]('apple');
        });
      });
    });

    describe('regexp', () => {
      it('passes', () => {
        expect(() => { throw new Error('apple'); })[toThrow](/apple/);
        expect(() => { throw new Error('banana'); }).not[toThrow](/apple/);
        expect(() => {}).not[toThrow](/apple/);
      });

      test('did not throw at all', () => {
        matchErrorSnapshot(() => jestExpect(() => {})[toThrow](/apple/));
      });

      test('threw, but message did not match', () => {
        matchErrorSnapshot(() => {
          jestExpect(() => { throw new Error('apple'); })
            [toThrow](/banana/);
        });
      });

      test('threw, but should not have', () => {
        matchErrorSnapshot(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow](/apple/);
        });
      });
    });

    describe('errors', () => {
      it('works', () => {
        it('passes', () => {
          jestExpect(() => { throw new Err(); })[toThrow](new Err());
          jestExpect(() => { throw new Err('Message'); })
            [toThrow](new Err('Message'));
          jestExpect(() => { throw new Err(); })[toThrow](new Error());
          jestExpect(() => { throw new Err(); }).not[toThrow](new Err2());
          jestExpect(() => {}).not[toThrow](new Err());
        });
      });
    });

    describe('error class', () => {
      it('passes', () => {
        jestExpect(() => { throw new Err(); })[toThrow](Err);
        jestExpect(() => { throw new Err(); })[toThrow](Error);
        jestExpect(() => { throw new Err(); }).not[toThrow](Err2);
        jestExpect(() => {}).not[toThrow](Err);
      });

      test('did not throw at all', () => {
        matchErrorSnapshot(() => expect(() => {})[toThrow](Err));
      });

      test('threw, but class did not match', () => {
        matchErrorSnapshot(() => {
          jestExpect(() => { throw new Err('apple'); })[toThrow](Err2);
        });
      });

      test('threw, but should not have', () => {
        matchErrorSnapshot(() => {
          jestExpect(() => { throw new Err('apple'); }).not[toThrow](Err);
        });
      });
    });

    test('invalid arguments', () => {
      matchErrorSnapshot(() => jestExpect(() => {})[toThrow](111));
    });
  });

});
