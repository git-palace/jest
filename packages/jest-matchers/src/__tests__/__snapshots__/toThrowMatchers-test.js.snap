exports[`.toThrowError() error class did not throw at all 1`] = `"Expected the function to throw an error of [36m[1m"Err"[22m[39m type, but it didn't."`;

exports[`.toThrowError() error class threw, but class did not match 1`] = `
"Expected the function to throw an error of [36m[1m"Err2"[22m[39m type, but it didn't.
Actual error:
  Type: [36m[1m"Err"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;

exports[`.toThrowError() error class threw, but should not have 1`] = `
"Expected the function to not throw an error of [36m[1m"Err"[22m[39m type, but it did.
Actual error:
  Type: [36m[1m"Err"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;

exports[`.toThrowError() regexp did not throw at all 1`] = `"Expected the function to throw an error matching [36m[1m"/apple/"[22m[39m, but it didn't."`;

exports[`.toThrowError() regexp threw, but message did not match 1`] = `
"Expected the function to throw an error matching [36m[1m"/banana/"[22m[39m, but it didn't.
Actual error:
  Type: [36m[1m"Error"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;

exports[`.toThrowError() regexp threw, but should not have 1`] = `
"Expected the function to not throw an error matching [36m[1m"/apple/"[22m[39m, but it did.
Actual error:
  Type: [36m[1m"Error"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;

exports[`.toThrowError() strings did not throw at all 1`] = `"Expected the function to throw an error matching [36m[1m"apple"[22m[39m, but it didn't."`;

exports[`.toThrowError() strings threw, but message did not match 1`] = `
"Expected the function to throw an error matching [36m[1m"banana"[22m[39m, but it didn't.
Actual error:
  Type: [36m[1m"Error"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;

exports[`.toThrowError() strings threw, but should not have 1`] = `
"Expected the function to not throw an error matching [36m[1m"apple"[22m[39m, but it did.
Actual error:
  Type: [36m[1m"Error"[22m[39m
  Message: [36m[1m"apple"[22m[39m"
`;
